File: src/routes/questions.py
from io import BytesIO

from flask_restplus import Resource, reqparse, abort, marshal_with, fields, Namespace
from werkzeug.datastructures import FileStorage

from chanakya.src import app, db
from chanakya.src.models import (
					Student,
					IncomingCalls,
					StudentContact,
					Questions
			)

from chanakya.src.helpers.response_objects import question_obj, questions_list_obj
from chanakya.src.helpers.task_helpers import render_pdf_phantomjs
from chanakya.src.helpers.file_uploader import upload_file_to_s3, FileStorageArgument
from chanakya.src.helpers.validators import check_option_ids

api = Namespace('questions', description='Handle CRUD and file upload related to questions for the Test')

@api.route('/upload_file')
class UploadQuestionImage(Resource):

	post_parser = reqparse.RequestParser(argument_class=FileStorageArgument)
	post_parser.add_argument('image', required=True, type=FileStorage, location='files')

	@api.doc(parser=post_parser)
	def post(self):

		args = self.post_parser.parse_args()

		# upload to s3
		image = args['image']
		image_url = upload_file_to_s3(app.config['S3_QUESTION_IMAGES_BUCKET'], image)

		return {'image_url': image_url}


@api.route('')
class QuestionList(Resource):

	get_response = api.model('GET_questions_list', {
		'data' : fields.List(fields.Nested(question_obj)),
		'error':fields.Boolean(default=False),
		'message':fields.String
	})

	# Description of POST method
	CREATE_QUESTION = """
	Possible values of different JSON keys which can be passed.

	- 'type': ['MCQ', 'Integer Answer'],
	- 'topic': ['BASIC_MATH', 'ABSTRACT_REASONING', 'NON_VERBAL_LOGICAL_REASONING'],
	- 'difficulty': ['Easy', 'Medium', 'Hard'],
	- 'en_text': Question string in English,
	- 'hi_text': Question string in Hindi,

	- 'options': This will contain an array of options. Every object in the array will look like:
		[
		    {
		        'en_text': 'Option in English',
		        'hi_text': 'Option in Hindi',
		        'correct': True  if it's correct option for the question else False
		    }
		]
	"""

	# create question
	post_model_option = api.model('POST_questions_options',{
	    "hi_text": fields.String(required=True),
	    "en_text": fields.String(required=True),
	    "correct": fields.Boolean(default=False, required=True)
	})

	post_payload_model = api.model('POST_questions',{
	    'en_text': fields.String(required=True),
	    'hi_text': fields.String(required=True),
	    'difficulty': fields.String(enum=[attr.value for attr in app.config['QUESTION_DIFFICULTY']], required=True),
	    'topic': fields.String(enum=[attr.value for attr in app.config['QUESTION_TOPIC']], required=True),
	    'type': fields.String(enum=[attr.value for attr in app.config['QUESTION_TYPE']], required=True),
	    'options': fields.List(fields.Nested(post_model_option), required=True)
	})

	post_response = api.model('POST_questions_response', {
		'error': fields.Boolean(default=False),
		'message':fields.String,
		'data': fields.Nested(question_obj)
	})

	@api.marshal_with(get_response)
	def get(self):
		questions_list = Questions.query.all()
		if questions_list:
			return {
				"data":questions_list
			}
		return {
			'error':True,
			'message':'There no question on platform. Please add some question'
		}

	@api.marshal_with(post_response)
	@api.expect(post_payload_model)
	@api.doc(description=CREATE_QUESTION)
	def post(self):
		args = api.payload
		# create the question
		options = args.get('options')
		is_any_option_correct = False
		for option in options:
			if option['correct']:
				is_any_option_correct = True

		if not is_any_option_correct:
			return{
				'error':True,
				'message': 'Minimum 1 option should be correct for the question.'
			}
		question = Questions.create_question(args)
		return {'data':question}


@api.route('/<question_id>')
class Question(Resource):

	get_response = api.model('GET_questions_id_response', {
		'error': fields.Boolean(default=False),
		'data': fields.Nested(question_obj),
		'message': fields.String
	})


	put_response = api.model('PUT_questions_id_response', {
		'error': fields.Boolean(default=False),
		'message': fields.String,
		'invalid_option_ids': fields.List(fields.Integer),
		'data' : fields.Nested(question_obj)
	})

	put_payload_model = question_obj

	QUESTIONS_UPDATE_DESCRIPTION = """
		Possible values of different JSON keys which can be passed:

			- 'id': Id of the Question in Integer,
			- 'type': ['MCQ', 'Integer Answer'],
			- 'topic': ['BASIC_MATH', 'ABSTRACT_REASONING', 'NON_VERBAL_LOGICAL_REASONING'],
			- 'difficulty': ['Easy', 'Medium', 'Hard'],
			- 'en_text': Question string in English,
			- 'hi_text': Question string in Hindi,

			- 'options': This will contain an array of options. Every object in the array will look like:
				[
				    {
						'id': Id of the Option in Integer,
				        'en_text': 'Option in English',
				        'hi_text': 'Option in Hindi',
				        'correct': True  if it's correct option for the question else False
				    }
				]

			RETURN fields:

			- 'invalid_option_ids' : [ 1, 4, 11, 112, 131, 142, 122] Invalid options ids.
	"""


	@api.marshal_with(get_response)
	def get(self, question_id):
		# return a single question
		question = Questions.query.filter_by(id = question_id).first()

		# find question or error
		if not question:
			return {
				'error': True,
				'message': "Question with given ID doesn't exist."
			}

		return { 'data': question }

	@api.doc(description=QUESTIONS_UPDATE_DESCRIPTION)
	@api.marshal_with(put_response)
	@api.expect(put_payload_model)
	def put(self, question_id):
		args = api.payload

		# find question or error
		question = Questions.query.filter_by(id = question_id).first()
		if not question:
			return {
				'error':True,
				'message': "Question id doesn't exist",
			}

		# In the PUT request the IDs of the options being edited need to be mentioned.
		# Check if any wrong option IDs are attached. If yes, then throw an error.
		wrong_option_ids = check_option_ids(question, args)
		if wrong_option_ids:
			return {
				'error':True,
				'message':'Incorrect option_id for the question',
				'invalid_option_ids': wrong_option_ids
			}

		# Update the question
		question.update_question(args)

		return { 'data': question }

File: lib/routes/questions.js
const Joi = require('joi');
const Question = require('../models/question');
const QuestionOption = require('../models/questionOption');
const EnrolmentKey = require('../models/enrolmentKey');
const logger = require('../../server/logger');

const internals = {};

// Joi schema for question Object
internals.questionSchema = Joi.object({
  common_text: Question.field('common_text'),
  en_text: Question.field('en_text'),
  hi_text: Question.field('hi_text'),
  ma_text: Question.field('ma_text'),
  difficulty: Question.field('difficulty'),
  topic: Question.field('topic'),
  type: Question.field('type'),
  schoolId: Joi.number().integer().allow(null).optional(),
}).options({ stripUnknown: true });

internals.questionOptionSchema = Joi.object({
  text: QuestionOption.field('text'),
  correct: QuestionOption.field('correct'),
});

module.exports = [
  {
    method: 'GET',
    path: '/questions',
    options: {
      description: 'Return a list of all the questions.',
      tags: ['api'],
      handler: async (request) => {
        const { questionService } = request.services();
        const questions = await questionService.findAll();
        logger.info('Return a list of all the questions.');
        return { data: questions };
      },
    },
  },
  {
    method: 'POST',
    path: '/questions',
    options: {
      description: 'Create a new question.',
      tags: ['api'],
      validate: {
        options: Joi.object({
          allowUnknown: true,
        }),
        payload: internals.questionSchema.concat(
          Joi.object({
            options: Joi.array().min(1).items(internals.questionOptionSchema).required(),
          })
        ),
      },
      handler: async (request) => {
        const { questionService } = request.services();
        // const question = await Promise.all(request.payload.map(question => questionService.create(question)));
        const question = await questionService.create(request.payload);
        // const question = await questionService.create(request.payload);
        logger.info('Create a new question.');
        return { data: question };
      },
    },
  },
  {
    method: 'GET',
    path: '/questions/{questionId}',
    options: {
      description: 'Get details of the question with the given ID.',
      tags: ['api'],
      validate: {
        params: Joi.object({
          questionId: Question.field('id'),
        }),
      },
      handler: async (request) => {
        const { questionService } = request.services();
        const question = await questionService.findById(request.params.questionId);
        logger.info('Get details of the question with the given ID.');
        return { data: question };
      },
    },
  },
  {
    method: 'PUT',
    path: '/questions/{questionId}',
    options: {
      description: 'Edit the question with a given ID.',
      tags: ['api'],
      validate: {
        params: Joi.object({
          questionId: Question.field('id'),
        }),
        payload: Joi.object({
          common_text: Question.field('common_text'),
          en_text: Question.field('en_text'),
          hi_text: Question.field('hi_text'),
          ma_text: Question.field('ma_text'),
          difficulty: Question.field('difficulty'),
          topic: Question.field('topic'),
          type: Question.field('type'),
          id: Question.field('id'),
          options: Joi.array(),
        }).options({ stripUnknown: true }),
      },
      handler: async (request) => {
        const { questionService } = request.services();
        const question = await questionService.update(request.payload);
        logger.info('Edit the question with a given ID.');
        return { data: question };
      },
    },
  },
  {
    method: 'DELETE',
    path: '/questions/{questionId}',
    options: {
      description: 'Delete the question with a given ID.',
      tags: ['api'],
      validate: {
        params: Joi.object({
          questionId: Question.field('id'),
        }),
      },
      handler: async (request) => {
        const { questionService } = request.services();
        await questionService.delete(request.params.questionId); // responce
        logger.info('Delete the question with a given ID.');
        return { success: true };
      },
    },
  },

  {
    method: 'GET',
    path: '/questions/attempts/{enrolmentKeyId}',
    options: {
      description: 'Get the questions by a given enrolmentKeyId.',
      tags: ['api'],
      validate: {
        params: Joi.object({
          enrolmentKeyId: EnrolmentKey.field('id'),
        }),
      },
      handler: async (request) => {
        const { questionAttemptService } = request.services();
        const attempts = await questionAttemptService.findQuestionIdByEnrolmentKeyId(
          request.params.enrolmentKeyId
        );
        logger.info('Get the questions by a given enrolmentKeyId.');
        return attempts;
      },
    },
  },
];
